### ISSUE TITLE
Find a more general type for mfix 

### ISSUE THREAD

**Issue by Lysxia** [DeepSpec/InteractionTrees#35]
I'm referring to this type: https://github.com/DeepSpec/InteractionTrees/blob/ea07080d9d7099e7e8418dab55fa6372e2926b2e/theories/MFixITree.v#L38

This type does not allow certain nestings of `mfix`, because the inner monad is too abstract. More precisely, the body of the inner `mfix` cannot make a call to the outer `mfix`. To illustrate, in OCaml syntax:

```ocaml
let rec f x =
  let rec g y =
    (* we should be able to call f here *)
  in ...
in ...
```

Note that the actual implementation of `mfix` for `itree` does not have this problem, thanks to `M` and `N` being instantiated with `itree`.

My current motivation for wanting as general a signature as possible for `mfix` is to instantiate it with a much more performant `IO` monad (I defined in coq-simple-io). The idea is to write programs as ``forall m `{Monad m} `{MonadFix m} ..., m unit``, which yields reasonably efficient extracted code by instantiating `m` with `IO`, and allows reasoning by instantiating `m` with some `itree E`. The current signature of `mfix` in InteractionTrees limits the expressiveness of this embedding.

**Comment by gmalecha** [comment]
Why can you not use `inc`? For example,

```coq
mfix (fun _ _ inc rec x =>
  mfix (fun _ _ inc' rec' y =>
   inc' (rec y) ...))
```

**Comment by Lysxia** [comment]
Unfortunately the instance of `mfix` for `itree` makes use of the fact that `inc` doesn't have to be surjective.

**Comment by gmalecha** [comment]
It doesn't really have to be an isomorphism, you can always interpret the new effect into the `mfix` construct itself, but in practice it probably won't work. I wonder if there is some way that we could prove that it isn't possible to write a more general `mfix` combinator?

**Comment by Lysxia** [comment]
`aloop` or `loop`, basically combinators for "tail recursion", seem general enough (as demonstrated by #79). I haven't thought this through yet, just triaging issues to see where we are at.

**Comment by Lysxia** [comment]
`aloop` seems to work. We can wrap in into a class `ALoop (M : Type -> Type) : Type`, such that

- there is an instance `forall E, ALoop (itree E)`
- we can interpret itrees into monads with `ALoop`, that is to say: `interp : Monad M -> ALoop M -> (E ~> M) -> (itree E ~> M)`
- we can write an instance `Monad M -> ALoop M -> ALoop (stateT S M)`, which gives us `interp_state` for free from `interp`, and same thing for `readerT` and `writerT`.

I've been tweaking the type of `aloop` a bit so I wouldn't need the `Monad M` constraints above, making the type signatures a bit simpler (i.e., `ALoop` itself contains the `Monad` structure). Interestingly, I ended up at a type which is exactly `itreeF`:

```coq
Class ALoop (M : Type -> Type) : Type :=
  aloop : forall I R, (I -> itreeF M R I) -> I -> M R.
```

Sadly I ran into some universe inconsistencies, I'm currently stuck in `FixFacts`, on the aloop-gen branch (based on the reorg PR): https://github.com/DeepSpec/InteractionTrees/commit/989ca75b2e78501d89d263b87836cfe0e02a0de2
